### PR TITLE
[Android] add Keep screen awake toggle

### LINKED ISSUE #2748: Android: Keep lights on
First, congratulations for releasing an Android version.

However, on Android it usually doesn't make sense to have a menu item that says E-ink optimization. It makes much better sense to have settings to adjust backlight and to block it from dimming. Then again, some e-ink devices have indeed Android.

### LINKED ISSUE #4364: Keep screen awake still not available
* KOReader version: 2018.11.1
* Device: Galaxy S7

#### Issue

I commented here https://github.com/koreader/koreader/issues/2748 that an option to keep a device screen from turning off is still not available despite a commit to to fix.

### PR DESCRIPTION
**Pull request by pazos**
continuation of #4409 , which I managed to close somehow, sorry :p 

Fixes #2748
Fixes #4364
Related to #4430 #3567 #3517 

In *some* phones *some* runtime configuration changes will force a surfaceDestroyed (followed by surfaceCreate) without losing/regain focus, so we need to refresh the screen on APP_CMD_INIT_WINDOW too. This will help on further "black screen issues"

More info: https://developer.nvidia.com/fixing-common-android-lifecycle-issues-games

### CODE REVIEW

**Comment by Frenzie** [comment]
You know you can reopen it, right? :-P

File: frontend/device/android/device.lua [hunk 1]
```diff
@@ -30,9 +30,9 @@ function Device:init()
             logger.dbg("Android application event", ev.code)
             if ev.code == C.APP_CMD_SAVE_STATE then
                 return "SaveState"
-            elseif ev.code == C.APP_CMD_GAINED_FOCUS then
-                this.device.screen:refreshFull()
-            elseif ev.code == C.APP_CMD_WINDOW_REDRAW_NEEDED then
+            elseif ev.code == C.APP_CMD_GAINED_FOCUS
+                or ev.code == C.APP_CMD_INIT_WINDOW
```
Comment (Frenzie) [inline frontend/device/android/device.lua]: I wonder if this might also fix the other issue with the brightness level.

Comment (pazos) [reply frontend/device/android/device.lua]: Hopefully :+1:

**Comment by pazos** [comment]
Uh oh, I guess that we need to restore the flag on init if `G_reader_settings:isTrue("keep_screen_on")` to allow starting w/ last saved state.

**Comment by pazos** [comment]
@Frenzie ready for review, again.

It would be nice to release this asap and see what happens with those crashy e-ink readers.

Do we need a wiki/doc entry for this change?

File: frontend/device/android/device.lua [hunk 2]
```diff
@@ -60,6 +60,11 @@ function Device:init()
         self.isTouchDevice = yes
     end
 
+    -- check if we disabled support for wakelocks
+    if G_reader_settings:isTrue("disable_wakelock") then
```
Comment (Frenzie) [inline frontend/device/android/device.lua]: Since the setting is Android-specific, should we make that clear in the setting name? Pinging @poire-z

Comment (poire-z) [reply frontend/device/android/device.lua]: I wouldn't mind, as only Android would get that setting. But `disable_android_wakelock` sounds explicite. Or `use_android_wakelocks=false`

Comment (Frenzie) [reply frontend/device/android/device.lua]: Yeah, let's go for `disable_android_wakelock`. :+1:

**Comment by rola25** [comment]
I own a Tolino Epos and could test the release when it is ready. I already tried it with "Wake Lock PowerManager", but this didn't work (still no refreshes in koreader) :-(

**Comment by Frenzie** [comment]
@pazos Please rebase & merge after the final change. :-)

**Comment by pazos** [comment]
@rola25:

You can help even when the device doesn't refresh at all (once this PR hits nightlies).

You will need adb installed and following the instructions

1. check if you can see your device with `adb devices`
2. clear old adb logs with `adb logcat -c`
3. uninstall KOReader app and install it again.
4. Launch KOReader, wait at least 15 seconds without touching the screen.
5. Do a few touches here and there. Wait another few seconds
6. Do a final touch.
7. generate a logfile with `adb logcat KOReader:V luajit-launcher:V *:S d > log.txt`
8. Share the log.txt file here.

Also I'm interested in getting a list of vendor/sytem libraries. Which can be done with

`adb shell 'find /system'` and `adb shell 'ls /vendor/lib'`

**Comment by rola25** [comment]
OK no prob. As soon as the .apk is available I will test it

**Comment by rola25** [comment]
In the meanwhile - here are the system files (no /vendor)

[system.txt](https://github.com/koreader/koreader/files/2748456/system.txt)

**Comment by rola25** [comment]
And here the logcat from the old version ...
[log.txt](https://github.com/koreader/koreader/files/2748534/log.txt)

**Comment by pazos** [comment]
@rola25: aweeeeesome, thanks!

Would you mind to share /system/lib/libepd.so? This is the electronic paper display library, used by some obscure frameworks to refresh the screen. I want to take a look.

BTW, the nightly with wakelocks support is [online](http://build.koreader.rocks/download/nightly/v2019.01.1-3-gab09ded_2019-01-11/). You can test it if you want 👍

**Comment by rola25** [comment]
Hi
Here ist libepd.so.
I tried the new nightly, but it still doesn't refresh. I also set the ["disable_android_wakelock"] = true option in settings.reader.lua.

Tomorrow I will do some more debugging and log-uploading.

[libepd.zip](https://github.com/koreader/koreader/files/2753166/libepd.zip)

**Comment by pazos** [comment]
@rola25: thanks again.

The good news are that your device is using an "standard" way of updating the screen: via `Java_de_telekom_epub_utils_ScreenHelper`

The bad news are that you need to wait a few weeks/months/years until somebody implements those methods in KOReader.

> Tomorrow I will do some more debugging and log-uploading.

If you wait a couple of days the log will be more valuable (it will provide your device model) and we can use that model to test further android e-ink aware implementations, since your device is currently unusable with ko.
 
```
00001f9d T Java_com_android_internal_policy_impl_PhoneWindowManager_FullRefresh
00002145 T Java_com_android_internal_policy_impl_PhoneWindowManager_getFBDoneRender
0000210d T Java_com_android_internal_policy_impl_PhoneWindowManager_setWantToCheckFBState
0000240d T Java_com_android_server_PowerManagerService_BlankFullRefresh
00001f9d T Java_com_android_server_PowerManagerService_FullRefresh
00002145 T Java_com_android_server_PowerManagerService_getFBDoneRender
00001f45 T Java_com_android_server_PowerManagerService_GetSleeping
00001eed T Java_com_android_server_PowerManagerService_SetSleeping
0000210d T Java_com_android_server_PowerManagerService_setWantToCheckFBState
00002b71 T Java_com_android_server_PowerManagerService_ShowScreen
00001f9d T Java_com_ntx_msg_MsgPowerOffActivity_FullRefresh
00001f45 T Java_com_ntx_msg_MsgSuspendActivity_GetSleeping
00001eed T Java_com_ntx_msg_MsgSuspendActivity_SetSleeping
00002f55 T Java_de_telekom_epub_utils_ScreenHelper_DrawBitmapToFB
00001ff5 T Java_de_telekom_epub_utils_ScreenHelper_FullRefresh
00001bc9 T Java_de_telekom_epub_utils_ScreenHelper_GetScreen
00001f45 T Java_de_telekom_epub_utils_ScreenHelper_GetSleeping
000020b3 T Java_de_telekom_epub_utils_ScreenHelper_PartialRefresh
00003081 T Java_de_telekom_epub_utils_ScreenHelper_PostFB
000020bf T Java_de_telekom_epub_utils_ScreenHelper_RegionalRefresh
00002411 T Java_de_telekom_epub_utils_ScreenHelper_SetFBColor
00002433 T Java_de_telekom_epub_utils_ScreenHelper_SetFBColorArray
00002481 T Java_de_telekom_epub_utils_ScreenHelper_SetUpdateMode
00001f45 T Java_de_telekom_test1_MainActivity_GetSleeping
00001eed T Java_de_telekom_test1_MainActivity_SetSleeping
00002f55 T Java_de_telekom_testeink_ImageActivity_DrawBitmapToFB
0000305d T Java_de_telekom_testeink_ImageActivity_PostFB
000020bf T Java_de_telekom_testeink_ImageActivity_RegionalRefresh
000020bf T Java_de_telekom_test_TestActivity_RegionalRefresh
00002411 T Java_de_telekom_test_TestActivity_SetFBColor
00002433 T Java_de_telekom_test_TestActivity_SetFBColorArray
0000240d T Java_ntx_epd_EPDRefreshActivity_BlankFullRefresh
00001f9d T Java_ntx_epd_EPDRefreshActivity_FullRefresh
00002145 T Java_ntx_epd_EPDRefreshActivity_getFBDoneRender
00001f45 T Java_ntx_epd_EPDRefreshActivity_GetSleeping
000020b3 T Java_ntx_epd_EPDRefreshActivity_PartialRefresh
000020bf T Java_ntx_epd_EPDRefreshActivity_RegionalRefresh
00001eed T Java_ntx_epd_EPDRefreshActivity_SetSleeping
0000210d T Java_ntx_epd_EPDRefreshActivity_setWantToCheckFBState
00001f9d T Java_ntx_power_AlarmReceiver_FullRefresh
00002f55 T Java_ntx_PowerEnhance_DemoPreload_DrawBitmapToFB
0000305d T Java_ntx_PowerEnhance_DemoPreload_PostFB
00002fbd T Java_ntx_power_PowerManagerEx_QuickPostFB
00001f9d T Java_ntx_power_PowerMechanismService_FullRefresh
00002145 T Java_ntx_power_PowerMechanismService_getFBDoneRender
0000210d T Java_ntx_power_PowerMechanismService_setWantToCheckFBStat
```

**Comment by rola25** [comment]
OK thanks for info and debugging :-) 
Just leave a comment here if/when you need new logs.

**Comment by NiLuJe** [comment]
Barring that, that looks like an NTX board, so the Kobo framebuffer code *could* be repurposed, but it's potentially a bit messy because I have no idea how we handle the framebuffer on Android ^^.

**Comment by pazos** [comment]
We cannot run code outside the VM on non rooted devices and even on rooted devices we will break [SurfaceFlinger](https://source.android.com/devices/graphics/arch-sf-hwc) if we try to write directly to the framebuffer a la Linux.

BTW a friend just lend me her Energy Pro 4 ereader for a few days. Device is based on RK chipset but hopefully the same logic can be applied to both freescale and rockchip devices. I will see what happens when I have a few spare hours :-)

**Comment by Frenzie** [comment]
@pazos Btw, you probably already saw, but just in case see this comment https://github.com/koreader/koreader/issues/3517#issuecomment-441762635 and the one right after https://github.com/koreader/koreader/issues/3517#issuecomment-445551470.

**Comment by NiLuJe** [comment]
@pazos : Eh, I was afraid of that ;).

So, yeah, that leaves the potentially-device-specific-and-vendor-provided JNI APIs...

**Comment by pazos** [comment]
@Frenzie: thanks for the reminder, AFAICT every open implementation is based on http://forum.xda-developers.com/showthread.php?t=1183173.

I will look at relaunchx first since seems easier to understand: https://github.com/rychly/relaunchx/blob/rychly/master/app/src/main/java/com/gacode/relaunchx/N2EpdController.java

**Comment by Frenzie** [comment]
@pazos Thanks for the links. I was not aware of that.

**Comment by pazos** [comment]
It seems that I've got the worst device to test. Tolino has an ScreenHelper class to handle libepd.so, Nook has the great N2EpdController and onyx devices have a full sdk w/ imx6 & rk3026 targets.

The reader I have just does `getWindow().getDecorView().invalidate(64526, 64526, 64526, 64526)` to force a full e-ink refresh. Oh, and boyue devices are a real crap, software wise, with a bunch of loggers and misterious activities. I doubt if they're using a floppy disk as the internal storage since my write/reads are 2MB/s & 10MB/s.

So, in short: in order to add full screen eink updates we need to add various epd controllers, a la Coolreader. It is not possible to probe current device epd controller without testing one by one (has a Nook epd?, has a Tolino epd?, has an Onyx epd? is a boyue crap?). This is bad because the ideal solution would be probing epd in the frontend. But instead of this we need to keep lists of "compatible" devices w/ each method and call that method directly.

I'm sad...

**Comment by Frenzie** [comment]
Sad, but not entirely unexpected based on coolreader. :-)
